File: typescript/streaming/src/shared/a2a-transformer.ts
/**
 * A2A stream transformer — emits A2A protocol-compliant streaming events.
 *
 * Uses the official `@a2a-js/sdk` types to ensure emitted events satisfy
 * `TaskStatusUpdateEvent` and `TaskArtifactUpdateEvent` shapes.
 *
 * Events are surfaced via a remote {@link StreamChannel} named `"a2a"`:
 *   - In-process consumers iterate `run.extensions.a2a` directly.
 *   - Remote SDK clients subscribe via `thread.subscribe("custom:a2a")`.
 */

import {
  AIMessage,
  AIMessageChunk,
} from "@langchain/core/messages";
import type { ProtocolEvent, StreamTransformer } from "@langchain/langgraph";
import { StreamChannel } from "@langchain/langgraph";
import type {
  TaskStatusUpdateEvent,
  TaskArtifactUpdateEvent,
} from "@a2a-js/sdk";

type A2AStreamEvent = TaskStatusUpdateEvent | TaskArtifactUpdateEvent;

/** LangGraph may emit `[payload, meta]` or a bare protocol dict for `messages`. */
function unpackMessagesPayload(raw: unknown): { payload: unknown } | null {
  if (Array.isArray(raw) && raw.length >= 2) {
    return { payload: raw[0] };
  }
  if (typeof raw === "object" && raw !== null && !Array.isArray(raw)) {
    return { payload: raw };
  }
  return null;
}

/**
 * Interpret LangGraph `messages` channel payloads where `event === "content-block-delta"` (`MessagesData`).
 *
 * Shape follows the **Agent Streaming Protocol** (`agent-protocol/streaming/protocol.cddl`;
 * generated TS in `agent-protocol/streaming/js/protocol.ts`). Each event carries `delta`:
 * **ContentBlockDelta** — append deltas (`text-delta`, `reasoning-delta`, `data-delta`) or
 * **block-delta**, which shallow-merges `fields` onto the active content block (tool-call argument
 * streaming and other merge semantics per the spec).
 *
 * A2A artifacts surface assistant-visible **text** (`text-delta`, plus legacy `{ type: "text", text }`
 * shapes); tool streams carried as `block-delta` (`tool_call_chunk`, `input_json_delta`, …) are skipped.
 */
function interpretContentBlockDelta(data: Record<string, unknown>): {
  skipArtifact: boolean;
  textChunk?: string;
} {
  const delta = data.delta as Record<string, unknown> | undefined;
  if (!delta || typeof delta !== "object") {
    return { skipArtifact: false };
  }

  if (delta.type === "block-delta") {
    const fields = delta.fields as Record<string, unknown> | undefined;
    if (!fields || typeof fields !== "object") {
      return { skipArtifact: false };
    }
    const ft = fields.type as string | undefined;
    if (
      ft === "tool_call_chunk" ||
      ft === "tool_use" ||
      ft === "tool_call" ||
      ft === "input_json_delta"
    ) {
      return { skipArtifact: true };
    }
    if (
      (ft === "text" || ft === "text-delta") &&
      typeof fields.text === "string"
    ) {
      return { skipArtifact: false, textChunk: fields.text };
    }
    return { skipArtifact: false };
  }

  const dt = delta.type as string | undefined;
  if (
    (dt === "text" || dt === "text-delta") &&
    typeof delta.text === "string"
  ) {
    return { skipArtifact: false, textChunk: delta.text };
  }

  return { skipArtifact: false };
}

export const createA2ATransformer = (): StreamTransformer<{
  a2a: StreamChannel<A2AStreamEvent>;
}> => {
  const a2a = StreamChannel.remote<A2AStreamEvent>("a2a");
  let started = false;

  let activeNode: string | undefined;
  let activeRole: string | undefined;
  let isToolCall = false;
  let accumulatedText = "";
  let artifactIndex = 0;

  /** Track which top-level subgraphs have been announced. */
  const announcedNodes = new Set<string>();

  const contextId = crypto.randomUUID();
  const taskId = crypto.randomUUID();

  const makeStatusEvent = (
    state: TaskStatusUpdateEvent["status"]["state"],
    text: string,
    final: boolean
  ): TaskStatusUpdateEvent => ({
    kind: "status-update",
    contextId,
    taskId,
    final,
    status: {
      state,
      message: {
        kind: "message",
        messageId: crypto.randomUUID(),
        role: "agent",
        parts: [{ kind: "text", text }],
      },
      timestamp: new Date().toISOString(),
    },
  });

  const makeArtifactEvent = (
    text: string,
    lastChunk: boolean
  ): TaskArtifactUpdateEvent => ({
    kind: "artifact-update",
    contextId,
    taskId,
    lastChunk,
    append: !lastChunk,
    artifact: {
      artifactId: `${activeNode ?? "agent"}-response-${artifactIndex}`,
      name: `${activeNode ?? "agent"}-response`,
      parts: [{ kind: "text", text }],
    },
  });

  return {
    init: () => ({ a2a }),

    process(event: ProtocolEvent) {
      if (!started) {
        started = true;
        a2a.push(
          makeStatusEvent("working", "Agent started processing", false)
        );
      }

      const ns = event.params.namespace ?? [];

      if (ns.length >= 1) {
        const segment = ns[0];
        const nodeName = segment.split(":")[0];
        if (!announcedNodes.has(nodeName)) {
          announcedNodes.add(nodeName);
          a2a.push(makeStatusEvent("working", `${nodeName} started`, false));
        }
      }

      if (event.method === "messages") {
        const unpacked = unpackMessagesPayload(event.params.data);
        if (!unpacked) {
          return true;
        }

        const { payload } = unpacked;

        if (AIMessageChunk.isInstance(payload)) {
          return true;
        }

        if (AIMessage.isInstance(payload)) {
          if (payload.tool_calls?.length) {
            return true;
          }
          const segment = ns[0] ?? "agent";
          activeNode = segment.split(":")[0];
          const text = payload.text.trim();
          if (text.length > 0) {
            accumulatedText = text;
            a2a.push(makeArtifactEvent(text, true));
            artifactIndex += 1;
          }
          accumulatedText = "";
          activeNode = undefined;
          activeRole = undefined;
          return true;
        }

        if (typeof payload !== "object" || payload === null) {
          return true;
        }

        const data = payload as Record<string, unknown>;

        if (data.event === "message-start") {
          const segment = ns[0] ?? "agent";
          activeNode = segment.split(":")[0];
          const roleRaw = data.role as string | undefined;
          activeRole =
            roleRaw ??
            (typeof data.tool_call_id === "string" ? "tool" : "ai");
          accumulatedText = "";
          isToolCall = false;
        }

        if (data.event === "content-block-start") {
          const cb = data.content as Record<string, unknown> | undefined;
          if (
            cb?.type === "tool_call_chunk" ||
            cb?.type === "tool_call" ||
            cb?.type === "tool_use"
          ) {
            isToolCall = true;
          }
        }

        if (
          activeRole === "ai" &&
          !isToolCall &&
          data.event === "content-block-delta"
        ) {
          const { skipArtifact, textChunk } = interpretContentBlockDelta(data);
          if (!skipArtifact && textChunk !== undefined) {
            accumulatedText += textChunk;
            a2a.push(makeArtifactEvent(textChunk, false));
          }
        }

        if (
          activeRole === "ai" &&
          data.event === "message-finish" &&
          accumulatedText.length > 0
        ) {
          a2a.push(makeArtifactEvent(accumulatedText, true));
          artifactIndex += 1;
          accumulatedText = "";
          activeNode = undefined;
          activeRole = undefined;
        }

        if (data.event === "message-finish") {
          isToolCall = false;
        }
      }

      return true;
    },

    finalize() {
      a2a.push(
        makeStatusEvent("completed", "Agent finished successfully", true)
      );
    },

    fail(err) {
      const message =
        typeof err === "object" &&
        err !== null &&
        "message" in err &&
        typeof (err as { message: unknown }).message === "string"
          ? (err as { message: string }).message
          : String(err);

      a2a.push(makeStatusEvent("failed", message, true));
    },
  };
};

File: python/streaming/src/shared/a2a_transformer.py
"""A2A stream transformer — emits A2A protocol-compliant streaming events.

Mirrors the JS `createA2ATransformer` (which uses `@a2a-js/sdk`
`TaskStatusUpdateEvent` / `TaskArtifactUpdateEvent` types) but emits
plain dicts in the same shape since the Python A2A SDK isn't a
dependency of this cookbook.

Events are surfaced via a `StreamChannel("a2a")`:
- In-process consumers iterate `run.extensions["a2a"]` directly.
- Remote SDK clients subscribe via the auto-forwarded `custom:a2a`
  channel.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from langgraph.stream import StreamChannel, StreamTransformer


def _unwrap_messages(event: dict[str, Any]) -> Any:
    """Unwrap `(payload, metadata)` from a messages event's data."""
    data = event.get("params", {}).get("data")
    if isinstance(data, tuple) and len(data) == 2:
        return data[0]
    return data


def _iso_now() -> str:
    return datetime.now(timezone.utc).isoformat()


class A2ATransformer(StreamTransformer):
    """Emit A2A protocol events as `messages` flow through the run."""

    def __init__(self, scope: tuple[str, ...] = ()) -> None:
        super().__init__(scope)
        self._channel: StreamChannel[dict[str, Any]] = StreamChannel("a2a")
        self._started = False
        self._active_node: str | None = None
        self._active_role: str | None = None
        self._is_tool_call = False
        self._accumulated_text = ""
        self._artifact_index = 0
        self._announced_nodes: set[str] = set()
        self._context_id = str(uuid.uuid4())
        self._task_id = str(uuid.uuid4())

    def init(self) -> dict[str, Any]:
        return {"a2a": self._channel}

    def _status_event(self, state: str, text: str, final: bool) -> dict[str, Any]:
        return {
            "kind": "status-update",
            "contextId": self._context_id,
            "taskId": self._task_id,
            "final": final,
            "status": {
                "state": state,
                "message": {
                    "kind": "message",
                    "messageId": str(uuid.uuid4()),
                    "role": "agent",
                    "parts": [{"kind": "text", "text": text}],
                },
                "timestamp": _iso_now(),
            },
        }

    def _artifact_event(self, text: str, last_chunk: bool) -> dict[str, Any]:
        node = self._active_node or "agent"
        return {
            "kind": "artifact-update",
            "contextId": self._context_id,
            "taskId": self._task_id,
            "lastChunk": last_chunk,
            "append": not last_chunk,
            "artifact": {
                "artifactId": f"{node}-response-{self._artifact_index}",
                "name": f"{node}-response",
                "parts": [{"kind": "text", "text": text}],
            },
        }

    def process(self, event: dict[str, Any]) -> bool:
        if not self._started:
            self._started = True
            self._channel.push(
                self._status_event("working", "Agent started processing", False)
            )

        namespace = event.get("params", {}).get("namespace") or []
        if namespace:
            segment = namespace[0]
            node_name = segment.split(":", 1)[0]
            if node_name not in self._announced_nodes:
                self._announced_nodes.add(node_name)
                self._channel.push(
                    self._status_event("working", f"{node_name} started", False)
                )

        if event.get("method") != "messages":
            return True

        payload = _unwrap_messages(event)
        if not isinstance(payload, dict):
            return True

        evt = payload.get("event")

        if evt == "message-start":
            segment = namespace[0] if namespace else "agent"
            self._active_node = segment.split(":", 1)[0]
            self._active_role = payload.get("role") or "ai"
            self._accumulated_text = ""
            self._is_tool_call = False

        elif evt == "content-block-start":
            content = payload.get("content") or {}
            if isinstance(content, dict) and content.get("type") in (
                "tool_call_chunk",
                "tool_call",
                "tool_use",
            ):
                self._is_tool_call = True

        elif evt == "content-block-delta":
            delta = payload.get("delta") or {}
            if (
                isinstance(delta, dict)
                and not self._is_tool_call
                and self._active_role == "ai"
                and delta.get("type") == "text-delta"
            ):
                chunk = delta.get("text") or ""
                if chunk:
                    self._accumulated_text += chunk
                    self._channel.push(self._artifact_event(chunk, False))

        elif evt == "message-finish":
            if self._active_role == "ai" and self._accumulated_text:
                self._channel.push(
                    self._artifact_event(self._accumulated_text, True)
                )
                self._artifact_index += 1
                self._accumulated_text = ""
                self._active_node = None
                self._active_role = None
            self._is_tool_call = False

        return True

    def finalize(self) -> None:
        self._channel.push(
            self._status_event("completed", "Agent finished successfully", True)
        )

    def fail(self, err: BaseException) -> None:
        self._channel.push(self._status_event("failed", str(err), True))
